### PR TITLE
Refactor message components to share functionality

### LINKED ISSUE #226: Factor out common HTML structures from message types
Much of the structure of messages is common. We should factor out the parts that can be reused across message types.

We should wait to do this until we've implemented the set of important message types. If we do it too early, we might end up with premature generalization which leads to wrong abstractions. As Sarah Mei [says](https://twitter.com/sarahmei/status/702289102620614656), "If you're not 100% sure you've got the right concept - leave that duplication. You'll have more information next time you touch that code."

### PR DESCRIPTION
**Pull request by linclark**
Fixes #226 

@bgrins, this is the one I was telling you about. Thank you!!

### CODE REVIEW

**Comment by bgrins** [comment]
Getting this error when filtering with the patch applied:

JavaScript error: resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/webconsole/new-console-output/selectors/messages.js, line 103: TypeError: can't convert null to object

**Comment by linclark** [comment]
That's weird, I don't get an error. What kind of filtering are you doing?

**Comment by bgrins** [comment]
This is text filtering on this page: https://content-security-policy.com/browser-test/.   Just typing random text into the filter box (note that severity filtering does work).

I also get this error with npm test:

```
1) ConsoleAPICall component: console.trace renders:
   Error: Expected 0 to be 3
    at assert (node_modules/expect/lib/assert.js:29:9)
    at Expectation.toBe (node_modules/expect/lib/Expectation.js:66:28)
    at Context.<anonymous> (new-console-output/test/components/console-api-call.test.js:90:33)

2) PageError component: renders:
   Error: Expected 'http://example.com/browser/devtools/client/webconsole/new-console-output/test/fixtures/stub-generato' to be 'test-tempfile.js:3:5'
    at assert (node_modules/expect/lib/assert.js:29:9)
    at Expectation.toBe (node_modules/expect/lib/Expectation.js:66:28)
    at Context.<anonymous> (new-console-output/test/components/page-error.test.js:31:33)
```

File: devtools/client/webconsole/new-console-output/components/message.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+};
+
+function Message(props) {
+  const {
+    messageId,
+    open,
+    source,
+    type,
+    level,
+    topLevelClasses: classes,
```
Comment (bgrins) [inline devtools/client/webconsole/new-console-output/components/message.js]: Is there a reason to rename the variable here?  Seems we could use `topLevelClasses` throughout this component as well to make searching for usage of it a little easier

File: devtools/client/webconsole/new-console-output/components/message.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    messageBody,
+    frame,
+    stacktrace,
+    onViewSourceInDebugger,
+    sourceMapService,
+    dispatch,
+  } = props;
+
+  classes.push("message");
```
Comment (bgrins) [inline devtools/client/webconsole/new-console-output/components/message.js]: Nit: you can do this all in one line:

`classes.push("message", source, type, level)`

**Review by bgrins — APPROVED** [review]
It's a little hard to review but I've looked through each commit individually and I agree generally with the idea.  The only thing I wonder is if this will somehow affect console.group

**Comment by bgrins** [comment]
> The only thing I wonder is if this will somehow affect console.group

Ping @nchevobbe as a heads up about this - I'm not sure how you are implementing grouped messages but this is changing how other messages get rendered.

**Comment by bgrins** [comment]
Really quick test of test_render_perf.html:

With PR:

```
1 INFO TEST-START | devtools/client/webconsole/new-console-output/test/chrome/test_render_perf.html
2 INFO took 0.201435 seconds to render messages
3 INFO took 0.00042499999999995455 seconds to search filter half the messages
```

Without PR: 

```
1 INFO TEST-START | devtools/client/webconsole/new-console-output/test/chrome/test_render_perf.html
2 INFO took 0.20686 seconds to render messages
3 INFO took 0.0003650000000000091 seconds to search filter half the messages
```

File: devtools/client/webconsole/new-console-output/components/message.js (hunk 2 above)
Comment (linclark) [reply devtools/client/webconsole/new-console-output/components/message.js]: Good call, fixed

File: devtools/client/webconsole/new-console-output/components/message.js (hunk 1 above)
Comment (linclark) [reply devtools/client/webconsole/new-console-output/components/message.js]: You're right, it would make searching easier. CHanged

**Comment by nchevobbe** [comment]
> The only thing I wonder is if this will somehow affect console.group

Actually it is affecting it, in a good way. I added a `Indent` component that display a div with the correct width, and I went through each message to add it. So here, I guess it will only makes my patch easier to read ( even if I think I'll have conflicts when rebasing this).

**Pull request merged**